src/Analyzers/Core/Analyzers/ForEachCast/AbstractForEachCastDiagnosticAnalyzer.cs

```diff
@@ -84,12 +76,36 @@ protected override void InitializeWorker(AnalysisContext context)
                 return;
 
             var (conversion, collectionElementType) = GetForEachInfo(semanticModel, node);
+            if (collectionElementType is null)
+                return;
 
             // Don't bother checking conversions that are problematic for other reasons.  The user will already have a
             // compiler error telling them the foreach is in error.
             if (!conversion.Exists)
                 return;
 
+            // Consider:
+            // public class C : IEnumerable<Match>
+            // {
+            //     public IEnumerator GetEnumerator() => null; // compiler picks this for the foreach loop.
+            //
+            //     IEnumerator<Match> IEnumerable<Match>.GetEnumerator() => null; // compiler doesn't use this.
+            // }
+
+            // This collection have GetEnumerator method that returns non-strongly-typed IEnumerator, but also implements strongly-typed IEnumerable<T> explicitly.
+            // In this case, the compiler chooses the non-strongly-typed GetEnumerator and adds explicit cast for `foreach (Match m in new C())`.
+            // This cast can fail if the collection is badly implemented such that the strongly-typed and non-strongly-typed GetEnumerator implemetations return different types.
+            // Given it's very rare that implementation can be bad, and to reduce false positives, we adjust collectionElementType for the case above to be `Match` instead of object.
+            if (collectionElementType.SpecialType == SpecialType.System_Object)
+            {
+                var ienumerableOfT = collectionType.AllInterfaces.SingleOrDefault(i => i.OriginalDefinition.SpecialType == SpecialType.System_Collections_Generic_IEnumerable_T);
```

Comment:
SingleOrDefault is not ok.  This will throw if a type impelments this interfaces multiple types.  For example `: IEnumerable<int>, IEnumerable<string>`.  Please add test.

Reply:
(note: this does occur with some real world types we've seen that both implement multiple times to simulate a variance relationship with nonvariant types).

Reply:
@CyrusNajmabadi Hmmm.. In that case should we allow `foreach` with any of types? That will require more work in the analyzer to check the conversion of all the candidate types in `IEnumerable<T>`s. Or could I use `FirstOrDefault` for now and open an issue to follow-up with proper handling for multiple `IEnumerable<T>`s?

Reply:
I agree, if someone is implementing IEnumerable<T> multiple times, *and* they are doing the legacy `object Current` thing, *and* someone is getting an unwanted foreach cast diagnostic here, we can revisit and make this code more complex at that time, not now.

Reply:
```suggestion
                var ienumerableOfT = collectionType.AllInterfaces.FirstOrDefault(i => i.OriginalDefinition.SpecialType == SpecialType.System_Collections_Generic_IEnumerable_T);
```